File: Sources/FloatingLyricCore/Auth/SpotifyAuth.swift
import Foundation

public final class SpotifyAuth: @unchecked Sendable {
    public static let scope = "user-read-playback-state user-modify-playback-state"
    private static let authorizeURL = URL(string: "https://accounts.spotify.com/authorize")!
    private static let tokenURL = URL(string: "https://accounts.spotify.com/api/token")!
    private static let refreshMarginSeconds: TimeInterval = 60

    private let clientID: String
    private let http: HTTPClient
    private let store: TokenStore
    private let now: () -> Date

    private var accessTokenValue: String?
    private var expiresAt: Date?

    public init(clientID: String,
                http: HTTPClient,
                store: TokenStore,
                now: @escaping () -> Date = Date.init) {
        self.clientID = clientID
        self.http = http
        self.store = store
        self.now = now
    }

    public var isLoggedIn: Bool { store.readRefreshToken() != nil }

    public func authorizationURL(challenge: String, state: String, redirectURI: String) -> URL {
        var components = URLComponents(url: Self.authorizeURL, resolvingAgainstBaseURL: false)!
        components.queryItems = [
            .init(name: "response_type", value: "code"),
            .init(name: "client_id", value: clientID),
            .init(name: "scope", value: Self.scope),
            .init(name: "redirect_uri", value: redirectURI),
            .init(name: "state", value: state),
            .init(name: "code_challenge_method", value: "S256"),
            .init(name: "code_challenge", value: challenge),
        ]
        return components.url!
    }

    public func exchange(code: String, verifier: String, redirectURI: String) async throws {
        try await requestToken(form: [
            "grant_type": "authorization_code",
            "code": code,
            "redirect_uri": redirectURI,
            "client_id": clientID,
            "code_verifier": verifier,
        ], failureError: .authCancelled)
    }

    public func accessToken() async throws -> String {
        if let token = accessTokenValue, let expiresAt,
           expiresAt.timeIntervalSince(now()) > Self.refreshMarginSeconds {
            return token
        }
        guard let refreshToken = store.readRefreshToken() else { throw AppError.notLoggedIn }
        try await requestToken(form: [
            "grant_type": "refresh_token",
            "refresh_token": refreshToken,
            "client_id": clientID,
        ], failureError: .sessionExpired)
        guard let token = accessTokenValue else { throw AppError.sessionExpired }
        return token
    }

    /// Drops the cached access token so the next `accessToken()` refreshes.
    public func invalidateAccessToken() {
        accessTokenValue = nil
        expiresAt = nil
    }

    public func logOut() {
        invalidateAccessToken()
        store.deleteRefreshToken()
    }

    /// Test seam: installs an access token without a network round-trip.
    public func setAccessTokenForTesting(_ token: String, expiresIn: TimeInterval) {
        accessTokenValue = token
        expiresAt = now().addingTimeInterval(expiresIn)
    }

    private struct TokenResponse: Decodable {
        let access_token: String
        let refresh_token: String?
        let expires_in: Int
    }

    private func requestToken(form: [String: String], failureError: AppError) async throws {
        var request = URLRequest(url: Self.tokenURL)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        request.httpBody = Data(formEncode(form).utf8)

        let response = try await http.send(request)
        guard (200..<300).contains(response.status),
              let decoded = try? JSONDecoder().decode(TokenResponse.self, from: response.body)
        else {
            if failureError == .sessionExpired { logOut() }
            throw failureError
        }

        accessTokenValue = decoded.access_token
        expiresAt = now().addingTimeInterval(TimeInterval(decoded.expires_in))
        if let refresh = decoded.refresh_token { store.writeRefreshToken(refresh) }
    }

    private func formEncode(_ form: [String: String]) -> String {
        var allowed = CharacterSet.alphanumerics
        allowed.insert(charactersIn: "-._~")
        return form.map { key, value in
            let k = key.addingPercentEncoding(withAllowedCharacters: allowed) ?? key
            let v = value.addingPercentEncoding(withAllowedCharacters: allowed) ?? value
            return "\(k)=\(v)"
        }.joined(separator: "&")
    }
}

File: windows/src/FloatingLyric.Core/SpotifyAuth.cs
using System.Text.Json;

namespace FloatingLyric.Core;

public sealed class SpotifyAuth(
    string clientId, IHttpClient http, ITokenStore store, Func<DateTime>? now = null)
{
    public const string Scope = "user-read-playback-state user-modify-playback-state";

    private const string AuthorizeUrl = "https://accounts.spotify.com/authorize";
    private const string TokenUrl = "https://accounts.spotify.com/api/token";
    private const double RefreshMarginSeconds = 60;

    private readonly Func<DateTime> _now = now ?? (() => DateTime.UtcNow);
    private string? _accessToken;
    private DateTime? _expiresAt;

    public bool IsLoggedIn => store.ReadRefreshToken() is not null;

    public Uri AuthorizationUrl(string challenge, string state, string redirectUri)
    {
        var query = new Dictionary<string, string>
        {
            ["response_type"] = "code",
            ["client_id"] = clientId,
            ["scope"] = Scope,
            ["redirect_uri"] = redirectUri,
            ["state"] = state,
            ["code_challenge_method"] = "S256",
            ["code_challenge"] = challenge,
        };
        var encoded = string.Join("&", query.Select(kv =>
            $"{Uri.EscapeDataString(kv.Key)}={Uri.EscapeDataString(kv.Value)}"));
        return new Uri($"{AuthorizeUrl}?{encoded}");
    }

    public Task ExchangeAsync(string code, string verifier, string redirectUri) =>
        RequestTokenAsync(new Dictionary<string, string>
        {
            ["grant_type"] = "authorization_code",
            ["code"] = code,
            ["redirect_uri"] = redirectUri,
            ["client_id"] = clientId,
            ["code_verifier"] = verifier,
        }, new AppError.AuthCancelled());

    /// <summary>The only route to a token. Refreshes with a minute to spare,
    /// so no caller anywhere thinks about expiry.</summary>
    public async Task<string> AccessTokenAsync()
    {
        if (_accessToken is { } token && _expiresAt is { } expiry &&
            (expiry - _now()).TotalSeconds > RefreshMarginSeconds)
            return token;

        if (store.ReadRefreshToken() is not { } refresh)
            throw new AppErrorException(new AppError.NotLoggedIn());

        await RequestTokenAsync(new Dictionary<string, string>
        {
            ["grant_type"] = "refresh_token",
            ["refresh_token"] = refresh,
            ["client_id"] = clientId,
        }, new AppError.SessionExpired());

        return _accessToken!;
    }

    /// <summary>Forces the next call to refresh — used after a 401.</summary>
    public void InvalidateAccessToken() => _expiresAt = null;

    public void LogOut()
    {
        _accessToken = null;
        _expiresAt = null;
        store.DeleteRefreshToken();
    }

    public void SetAccessTokenForTesting(string token, double expiresIn)
    {
        _accessToken = token;
        _expiresAt = _now().AddSeconds(expiresIn);
    }

    private async Task RequestTokenAsync(Dictionary<string, string> form, AppError failure)
    {
        using var request = new HttpRequestMessage(HttpMethod.Post, TokenUrl)
        {
            Content = new FormUrlEncodedContent(form),
        };
        var response = await http.SendAsync(request);

        if (response.Status is < 200 or >= 300)
        {
            // A refresh that fails is a dead session: clear it rather than
            // retrying forever with a token Spotify has rejected.
            if (form["grant_type"] == "refresh_token") store.DeleteRefreshToken();
            throw new AppErrorException(failure);
        }

        JsonDocument json;
        try { json = JsonDocument.Parse(response.Body); }
        catch (JsonException) { throw new AppErrorException(failure); }

        using var _ = json;
        var root = json.RootElement;
        if (root.ValueKind != JsonValueKind.Object ||
            !root.TryGetProperty("access_token", out var accessToken))
            throw new AppErrorException(failure);

        _accessToken = accessToken.GetString();
        var expiresIn = root.TryGetProperty("expires_in", out var e) ? e.GetDouble() : 3600;
        _expiresAt = _now().AddSeconds(expiresIn);

        if (root.TryGetProperty("refresh_token", out var refresh) &&
            refresh.GetString() is { Length: > 0 } value)
            store.WriteRefreshToken(value);
    }
}
